File: Archived/2022/OEPE/0122.LeetCode.21.MergeTwoSortedLists/source.go
package leetcode

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil {
		return list2
	}
	if list2 == nil {
		return list1
	}

	merge := func(a, b *ListNode) *ListNode {
		a.Next = mergeTwoLists(a.Next, b)
		return a
	}

	if list1.Val <= list2.Val {
		return merge(list1, list2)
	} else {
		return merge(list2, list1)
	}
}

File: Archived/2022/OEPE/0122.LeetCode.21.MergeTwoSortedLists/source.py
from typing import *

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 is None:
            return list2
        if list2 is None:
            return list1
            
        head, tail = None, None        

        while list1 is not None and list2 is not None:
            if list1.val <= list2.val:
                if head is None:
                    head = list1                
                else:
                    tail.next = list1
                tail = list1    
                list1 = list1.next
            else:
                if head is None:
                    head = list2                    
                else:
                    tail.next = list2
                tail = list2
                list2 = list2.next

        if list1 is not None:            
            tail.next = list1
        if tail is not None and list2 is not None:
            tail.next = list2
        
        return head